src/common/nvtx.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef MXNET_COMMON_NVTX_H_
+#define MXNET_COMMON_NVTX_H_
+
+#if MXNET_USE_CUDA && MXNET_USE_NVTX
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <nvToolsExtCuda.h>
+#include <vector>
+#include <string>
+#include <cstring>
+
+namespace mxnet {
+namespace common {
+namespace cuda {
+
+class NVTXDuration {
+ public:
+  explicit NVTXDuration(const char *name) noexcept
+      : range_id_(0), name_(name) {}
+
+  inline void start() {
+    range_id_ = nvtxRangeStartA(name_);
+  }
+
+  inline void stop() {
+    nvtxRangeEnd(range_id_);
+  }
+
+ private:
+  nvtxRangeId_t range_id_;
+  const char *name_;
+};
+
+// Utility class for NVTX
+class nvtx {
+ public:
+  // Palette of colors (make sure to add new colors to the vector in nameToColor()).
+  static const uint32_t kRed     = 0xFF0000;
+  static const uint32_t kGreen   = 0x00FF00;
+  static const uint32_t kBlue    = 0x0000FF;
+  static const uint32_t kYellow  = 0xB58900;
+  static const uint32_t kOrange  = 0xCB4B16;
+  static const uint32_t kRed1    = 0xDC322F;
+  static const uint32_t kMagenta = 0xD33682;
+  static const uint32_t kViolet  = 0x6C71C4;
+  static const uint32_t kBlue1   = 0x268BD2;
+  static const uint32_t kCyan    = 0x2AA198;
+  static const uint32_t kGreen1  = 0x859900;
+
+  static void gpuRangeStart(const uint32_t rgb, const std::string& range_name) {
```

Comment:
Do we know if the nvtxRangePushEx(&att) makes a copy of att.message.ascii?  I suspect not, in which case I'm worried about the lifetime of range_name.c_str().

gpuRangeStart() is called with a const char *, so won't that require making a short-lived temporary string (with a copy of the char array) in the caller's environment?

Reply:
Not sure I your concern - nvtxRangePush is not an async call. It just puts the name and a timestamp into the database and that's it (and yes, it copies the string in the process).

Reply:
So if nvtxRangePush copies the att.message.ascii string (as opposed to the pointer), that answers my primary concern here.

I'm left with one nit then, why don't you have `const char * range_name` above, which avoids the needless string copy and keeps the type consistent between the callers (in c_api.cc and threaded_engine_perdevice.cc) and what att.message.ascii wants?

Reply:
It is a reference to a string, so there is no copying here (`c_str()` does not copy either). I believe in the threaded engine the nvtx_name is a string anyway, so that is also not doing any copies.

Reply:
In threaded engine, the following does a copy that could be avoided if everything was handled as a `const char *`:
```
auto nvtx_name = opr_block->opr->opr_name != "" ? opr_block->opr->opr_name : "Op";
```
so perhaps like:
```
auto nvtx_name = opr_block->opr->opr_name.size() ? opr_block->opr->opr_name.c_str() : "Op";
```
It's a nit as I say, so I'm OK leaving it as: `const std::string& range_name`.  Given that though, the two overloaded nameToColor() signatures should be consistent in their use of std::string (see an additional suggestion).